File: src/test/java/com/github/jactor/persistence/entity/PersistentEntityTest.java
package com.github.jactor.persistence.entity;

import static com.github.jactor.persistence.entity.AddressEntity.anAddress;
import static com.github.jactor.persistence.entity.BlogEntity.aBlog;
import static com.github.jactor.persistence.entity.BlogEntryEntity.aBlogEntry;
import static com.github.jactor.persistence.entity.GuestBookEntity.aGuestBook;
import static com.github.jactor.persistence.entity.GuestBookEntryEntity.aGuestBookEntry;
import static com.github.jactor.persistence.entity.PersonEntity.aPerson;
import static com.github.jactor.persistence.entity.UserEntity.aUser;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import com.github.jactor.persistence.dto.AddressInternalDto;
import com.github.jactor.persistence.dto.BlogDto;
import com.github.jactor.persistence.dto.BlogEntryDto;
import com.github.jactor.persistence.dto.GuestBookDto;
import com.github.jactor.persistence.dto.GuestBookEntryDto;
import com.github.jactor.persistence.dto.PersonInternalDto;
import com.github.jactor.persistence.dto.UserInternalDto;
import java.util.HashSet;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

@DisplayName("A PersistentEntity")
class PersistentEntityTest {

  private PersistentEntity<?> persistentEntityToTest;

  @Test
  @DisplayName("should be able to copy an address without the id")
  void shouldCopyAddress() {
    persistentEntityToTest = anAddress(
        new AddressInternalDto(null, "1001", "somewhere", "out", "there", "svg", "NO")
    );
    persistentEntityToTest.setId(1L);

    PersistentEntity<?> copy = (PersistentEntity<?>) persistentEntityToTest.copyWithoutId();

    assertAll(
        () -> assertThat(persistentEntityToTest).as("persistent entity").isNotNull(),
        () -> assertThat(persistentEntityToTest.getId()).as("id of persistent entity").isEqualTo(1L),
        () -> assertThat(copy).as("copy").isNotNull(),
        () -> assertThat(copy.getId()).as("id of copy").isNull(),
        () -> assertThat(persistentEntityToTest).as("persistent entity equals copy").isEqualTo(copy),
        () -> assertThat(persistentEntityToTest).as("persistent entity is not same instance as copy").isNotSameAs(copy)
    );
  }

  @Test
  @DisplayName("should be able to copy a person without the id")
  void shouldCopyPerson() {
    persistentEntityToTest = aPerson(
        new PersonInternalDto(null, new AddressInternalDto(), "us_US", "Bill", "Smith", "here i am")
    );
    persistentEntityToTest.setId(1L);

    PersistentEntity copy = (PersistentEntity<?>) persistentEntityToTest.copyWithoutId();

    assertAll(
        () -> assertThat(persistentEntityToTest).as("persistent entity").isNotNull(),
        () -> assertThat(persistentEntityToTest.getId()).as("id of persistent entity").isEqualTo(1L),
        () -> assertThat(copy).as("copy").isNotNull(),
        () -> assertThat(copy.getId()).as("id of copy").isNull(),
        () -> assertThat(persistentEntityToTest).as("persistent entity equals copy").isEqualTo(copy),
        () -> assertThat(persistentEntityToTest).as("persistent entity is not same instance as copy").isNotSameAs(copy)
    );
  }

  @Test
  @DisplayName("should be able to copy a user without the id")
  void shouldCopyUser() {
    persistentEntityToTest = aUser(new UserInternalDto(null, null, "i.am@home", "jactor"));
    persistentEntityToTest.setId(1L);

    PersistentEntity copy = (PersistentEntity<?>) persistentEntityToTest.copyWithoutId();

    assertAll(
        () -> assertThat(persistentEntityToTest).as("persistent entity").isNotNull(),
        () -> assertThat(persistentEntityToTest.getId()).as("id of persistent entity").isEqualTo(1L),
        () -> assertThat(copy).as("copy").isNotNull(),
        () -> assertThat(copy.getId()).as("id of copy").isNull(),
        () -> assertThat(persistentEntityToTest).as("persistent entity equals copy").isEqualTo(copy),
        () -> assertThat(persistentEntityToTest).as("persistent entity is not same instance as copy").isNotSameAs(copy)
    );
  }

  @Test
  @DisplayName("should be able to copy a blog without the id")
  void shouldCopyBlog() {
    persistentEntityToTest = aBlog(new BlogDto(null, null, "general ignorance", new UserInternalDto()));
    persistentEntityToTest.setId(1L);

    PersistentEntity copy = (PersistentEntity<?>) persistentEntityToTest.copyWithoutId();

    assertAll(
        () -> assertThat(persistentEntityToTest).as("persistent entity").isNotNull(),
        () -> assertThat(persistentEntityToTest.getId()).as("id of persistent entity").isEqualTo(1L),
        () -> assertThat(copy).as("copy").isNotNull(),
        () -> assertThat(copy.getId()).as("id of copy").isNull(),
        () -> assertThat(persistentEntityToTest).as("persistent entity equals copy").isEqualTo(copy),
        () -> assertThat(persistentEntityToTest).as("persistent entity is not same instance as copy").isNotSameAs(copy)
    );
  }

  @Test
  @DisplayName("should be able to copy a blog entry without the id")
  void shouldCopyBlogEntry() {
    BlogEntryDto blogEntryDto = new BlogEntryDto(null, new BlogDto(), "jactor", "the one");
    persistentEntityToTest = aBlogEntry(blogEntryDto);
    persistentEntityToTest.setId(1L);

    PersistentEntity copy = (PersistentEntity<?>) persistentEntityToTest.copyWithoutId();

    assertAll(
        () -> assertThat(persistentEntityToTest).as("persistent entity").isNotNull(),
        () -> assertThat(persistentEntityToTest.getId()).as("id of persistent entity").isEqualTo(1L),
        () -> assertThat(copy).as("copy").isNotNull(),
        () -> assertThat(copy.getId()).as("id of copy").isNull(),
        () -> assertThat(persistentEntityToTest).as("persistent entity equals copy").isEqualTo(copy),
        () -> assertThat(persistentEntityToTest).as("persistent entity is not same instance as copy").isNotSameAs(copy)
    );
  }

  @Test
  @DisplayName("should be able to copy a guest book without the id")
  void shouldCopyGuestBook() {
    persistentEntityToTest = aGuestBook(new GuestBookDto(null, new HashSet<>(), "enter when applied", new UserInternalDto()));
    persistentEntityToTest.setId(1L);

    PersistentEntity copy = (PersistentEntity<?>) persistentEntityToTest.copyWithoutId();

    assertAll(
        () -> assertThat(persistentEntityToTest).as("persistent entity").isNotNull(),
        () -> assertThat(persistentEntityToTest.getId()).as("id of persistent entity").isEqualTo(1L),
        () -> assertThat(copy).as("copy").isNotNull(),
        () -> assertThat(copy.getId()).as("id of copy").isNull(),
        () -> assertThat(persistentEntityToTest).as("persistent entity equals copy").isEqualTo(copy),
        () -> assertThat(persistentEntityToTest).as("persistent entity is not same instance as copy").isNotSameAs(copy)
    );
  }

  @Test
  @DisplayName("should be able to copy a guest book entry without the id")
  void shouldCopyGuestBookEntry() {
    persistentEntityToTest = aGuestBookEntry(
        new GuestBookEntryDto(null, new GuestBookDto(), "jactor", "the one"));
    persistentEntityToTest.setId(1L);

    PersistentEntity copy = (PersistentEntity<?>) persistentEntityToTest.copyWithoutId();

    assertAll(
        () -> assertThat(persistentEntityToTest).as("persistent entity").isNotNull(),
        () -> assertThat(persistentEntityToTest.getId()).as("id of persistent entity").isEqualTo(1L),
        () -> assertThat(copy).as("copy").isNotNull(),
        () -> assertThat(copy.getId()).as("id of copy").isNull(),
        () -> assertThat(persistentEntityToTest).as("persistent entity equals copy").isEqualTo(copy),
        () -> assertThat(persistentEntityToTest).as("persistent entity is not same instance as copy").isNotSameAs(copy)
    );
  }
}

File: src/test/kotlin/com/github/jactor/persistence/entity/PersistentEntityTest.kt
package com.github.jactor.persistence.entity

import java.util.UUID
import org.junit.jupiter.api.Test
import com.github.jactor.persistence.dto.AddressModel
import com.github.jactor.persistence.dto.BlogModel
import com.github.jactor.persistence.dto.BlogEntryModel
import com.github.jactor.persistence.dto.GuestBookModel
import com.github.jactor.persistence.dto.GuestBookEntryModel
import com.github.jactor.persistence.dto.PersistentDto
import com.github.jactor.persistence.dto.PersonModel
import com.github.jactor.persistence.dto.UserModel
import assertk.assertAll
import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isNotNull
import assertk.assertions.isNotSameInstanceAs
import assertk.assertions.isNull

internal class PersistentEntityTest {
    private lateinit var persistentEntityToTest: PersistentEntity<*>

    @Test
    fun `should be able to copy an address without the id`() {
        persistentEntityToTest = AddressBuilder.new(
            addressModel = AddressModel(
                persistentDto = PersistentDto(),
                zipCode = "1001",
                addressLine1 = "somewhere",
                addressLine2 = "out",
                addressLine3 = "there",
                city = "svg",
                country = "NO"
            )
        ).build()

        persistentEntityToTest.id = UUID.randomUUID()

        val copy = persistentEntityToTest.copyWithoutId() as PersistentEntity<*>

        assertAll {
            assertThat(persistentEntityToTest.id).isNotNull()
            assertThat(copy.id).isNull()
            assertThat(persistentEntityToTest).isEqualTo(copy)
            assertThat(persistentEntityToTest).isNotSameInstanceAs(copy)
        }
    }

    @Test
    fun `should be able to copy a person without the id`() {
        persistentEntityToTest = PersonBuilder.new(
            PersonModel(
                address = AddressModel(),
                locale = "us_US",
                firstName = "Bill",
                surname = "Smith", description = "here i am"
            )
        ).build()

        val copy = persistentEntityToTest.copyWithoutId() as PersistentEntity<*>

        assertAll {
            assertThat(persistentEntityToTest.id).isNotNull()
            assertThat(copy.id).isNull()
            assertThat(persistentEntityToTest).isEqualTo(copy)
            assertThat(persistentEntityToTest).isNotSameInstanceAs(copy)
        }
    }

    @Test
    fun `should be able to copy a user without the id`() {
        persistentEntityToTest = UserBuilder.new(
            UserModel(persistentDto = PersistentDto(), emailAddress = "i.am@home", username = "jactor")
        ).build()

        val copy = persistentEntityToTest.copyWithoutId() as PersistentEntity<*>

        assertAll {
            assertThat(persistentEntityToTest.id).isNotNull()
            assertThat(copy.id).isNull()
            assertThat(persistentEntityToTest).isEqualTo(copy)
            assertThat(persistentEntityToTest).isNotSameInstanceAs(copy)
        }
    }

    @Test
    fun `should be able to copy a blog without the id`() {
        persistentEntityToTest = BlogBuilder.new(
            blogModel = BlogModel(
                title = "general ignorance",
                userInternal = UserModel()
            )
        ).buildBlogEntity()

        val copy = persistentEntityToTest.copyWithoutId() as PersistentEntity<*>

        assertAll {
            assertThat(persistentEntityToTest.id).isNotNull()
            assertThat(copy.id).isNull()
            assertThat(persistentEntityToTest).isEqualTo(copy)
            assertThat(persistentEntityToTest).isNotSameInstanceAs(copy)
        }
    }

    @Test
    fun `should be able to copy a blog entry without the id`() {
        val blogEntryModel = BlogEntryModel(PersistentDto(), BlogModel(), "jactor", "the one")
        persistentEntityToTest = BlogBuilder.new().withEntry(blogEntryModel = blogEntryModel).buildBlogEntryEntity()
        val copy = persistentEntityToTest.copyWithoutId() as PersistentEntity<*>

        assertAll {
            assertThat(persistentEntityToTest.id).isNotNull()
            assertThat(copy.id).isNull()
            assertThat(persistentEntityToTest).isEqualTo(copy)
            assertThat(persistentEntityToTest).isNotSameInstanceAs(copy)
        }
    }

    @Test
    fun `should be able to copy a guest book without the id`() {
        persistentEntityToTest = GuestBookBuilder.new(
            guestBookModel = GuestBookModel(PersistentDto(), HashSet(), "enter when applied", UserModel())
        ).buildGuestBookEntity()

        val copy = persistentEntityToTest.copyWithoutId() as PersistentEntity<*>

        assertAll {
            assertThat(persistentEntityToTest.id).isNotNull()
            assertThat(copy.id).isNull()
            assertThat(persistentEntityToTest).isEqualTo(copy)
            assertThat(persistentEntityToTest).isNotSameInstanceAs(copy)
        }
    }

    @Test
    fun `should be able to copy a guest book entry without the id`() {
        persistentEntityToTest = GuestBookBuilder.new().withEntry(
            GuestBookEntryModel(
                persistentDto = PersistentDto(), guestBook = GuestBookModel(), creatorName = "jactor", entry = "the one"
            )
        ).buildGuestBookEntryEntity()

        val copy = persistentEntityToTest.copyWithoutId() as PersistentEntity<*>

        assertAll {
            assertThat(persistentEntityToTest.id).isNotNull()
            assertThat(copy.id).isNull()
            assertThat(persistentEntityToTest).isEqualTo(copy)
            assertThat(persistentEntityToTest).isNotSameInstanceAs(copy)
        }
    }
}
